src/storage.rs

```diff
@@ -61,40 +61,49 @@ pub struct MmapArray {
 }
 
 impl MmapChunk for MmapArray {
-    fn mmap_chunk(read: &mut BufReader<File>) -> Result<Self, Error> {
-        ensure!(
-            read.read_u32::<LittleEndian>()? == ChunkIdentifier::NdArray as u32,
-            "invalid chunk identifier for NdArray"
-        );
+    fn mmap_chunk(read: &mut BufReader<File>) -> Result<Self> {
+        ChunkIdentifier::check_read(read, ChunkIdentifier::NdArray)?;
 
         // Read and discard chunk length.
-        read.read_u64::<LittleEndian>()?;
-
-        let rows = read.read_u64::<LittleEndian>()? as usize;
-        let cols = read.read_u32::<LittleEndian>()? as usize;
+        read.read_u64::<LittleEndian>()
+            .map_err(|e| ErrorKind::io_error("Cannot read embedding matrix chunk length", e))?;
+
+        let rows = read.read_u64::<LittleEndian>().map_err(|e| {
+            ErrorKind::io_error("Cannot read number of rows of the embedding matrix", e)
+        })? as usize;
+        let cols = read.read_u32::<LittleEndian>().map_err(|e| {
+            ErrorKind::io_error("Cannot read number of columns of the embedding matrix", e)
+        })? as usize;
         let shape = Ix2(rows, cols);
 
-        ensure!(
-            read.read_u32::<LittleEndian>()? == f32::type_id(),
-            "Expected single precision floating point matrix for NdArray."
-        );
+        f32::check_read(read)?;
```

Comment:
Without context this is not self-explaining. Could be cleared up with some inline comment, that this line verifies the datatype for the embedding matrix.